internal/controllers/crds/dynamic_controller.go

```diff
@@ -0,0 +1,134 @@
+package crds
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/samber/lo"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/controllers/utils"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/util"
+)
+
+// +kubebuilder:rbac:groups="apiextensions.k8s.io",resources=customresourcedefinitions,verbs=list;watch
+
+type Controller interface {
+	SetupWithManager(mgr ctrl.Manager) error
+}
+
+// DynamicController ensures that RequiredCRDs are installed in the cluster and only then sets up all of its Controllers
+// that depends on them.
+// In case the CRDs are not installed at start-up time, DynamicController will set up a watch for CustomResourceDefinition
+// and will dynamically set up its Controllers once it detects that all RequiredCRDs are already in place.
+type DynamicController struct {
```

Comment:
How about 

```suggestion
type DynamicCRDController struct {
```

Or simply 


```suggestion
type CRDController struct {
```

?